src/Microsoft.ML.Data/Evaluators/Metrics/ConfusionMatrix.cs

```diff
@@ -0,0 +1,79 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.ML.Runtime;
+
+namespace Microsoft.ML.Data
+{
+    /// <summary>
+    /// Represents the <a href="https://en.wikipedia.org/wiki/Confusion_matrix">confusion matrix</a> of the classification results.
+    /// </summary>
+    public sealed class ConfusionMatrix
+    {
+        /// <summary>
+        /// The calculated value of <a href="https://en.wikipedia.org/wiki/Precision_and_recall#Precision">precision</a> for each class.
+        /// </summary>
+        public ImmutableArray<double> PerClassPrecision { get; }
```

Comment:
>ImmutableArray [](start = 15, length = 14)

That's fine, note that elsewhere in similar situations we've been using `IReadOnlyList`, but I guess this works too?

Reply:
I would also prefer to expose IReadOnlyList just for the sake of consistency. (I don't think we use ImmutableArray in public interface in metrics at all).

---
In reply to: [275005362](https://github.com/dotnet/machinelearning/pull/3250#discussion_r275005362) [](ancestors = 275005362)

Reply:
So @sfilipi I notice this was marked as resolved, if you disagree with me and @Ivanidzo4ka that's fine, but would you mind sharing justification? It is somewhat inconsistent?

---
In reply to: [275082837](https://github.com/dotnet/machinelearning/pull/3250#discussion_r275082837) [](ancestors = 275082837,275005362)

Reply:
Ivan has commented after i marked it as resolved, and i marked it as resolved because i thought your comment was just an observation, not a suggestion for change. I'll swap it. 

---
In reply to: [275472667](https://github.com/dotnet/machinelearning/pull/3250#discussion_r275472667) [](ancestors = 275472667,275082837,275005362)

Reply:
Did the change. Just wanted to mention that the reason i used arrays in the first place, is because i believe arrays are preferred over lists when it comes to small collections that don't get updated frequently. 

---
In reply to: [275504885](https://github.com/dotnet/machinelearning/pull/3250#discussion_r275504885) [](ancestors = 275504885,275472667,275082837,275005362)

Reply:
Thanks @sfilipi. You may be right, but for things like this I feel like using the more "consistent" solution with what we've done elsewhere, may be more desirable. (Even if one or two places we do use immutable arrays. See also `VectorDataViewType.Dimensions`, but that one was deliberate.)

It may be that in future we may decide using immutable arrays is more desirable. Incidentally this is allowed by [the rules](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md) is "Returning a value of a more derived type for a property, field, return or out value", so in principle we *could* change what we think of as read only lists to immutable arrays.

---
In reply to: [275554556](https://github.com/dotnet/machinelearning/pull/3250#discussion_r275554556) [](ancestors = 275554556,275504885,275472667,275082837,275005362)